# Introduction

We have a weather dataset of 26 North American cities from roughly 1960 to 2015, downloaded from the Global Climatology Network (https://www.ncdc.noaa.gov/data-access/land-based-station-data/land-based-datasets/global-historical-climatology-network-ghcn). The dataset consists of the following monthly weather data for each city: 
    1. average maximum temperature(0.1°C)
    2. average minimum temperature(0.1°C)
    3. average precipitation(0.1mm)
    4. average snowfall(mm)
    5. average snow depth(mm)
    
Using Python and relevant data science libraries, We plan to build and train an accurate machine learning algorithm to classify given unlabelled weather data into their corresponding North American cities. We will try a few different algorithms, validate them, and then choose the one with the best results.

# Set-up

We need to import Python's data science libraries and read the data file.

In [1]:
import numpy as np
import pandas as pd

data = pd.read_csv('monthly-data-labelled.csv')
#This is what the data looks like
data

city  year     tmax-01     tmax-02    tmax-03     tmax-04  \
0     Anchorage  1960  -46.516129   -9.482759  -9.677419   52.400000   
1     Anchorage  1961  -26.096774  -44.571429 -35.064516   58.200000   
2     Anchorage  1962  -59.225806  -31.750000 -18.903226   69.366667   
3     Anchorage  1963  -39.290323  -11.357143  -1.451613   41.700000   
4     Anchorage  1964  -59.129032  -24.655172 -35.096774   45.866667   
5     Anchorage  1965  -81.193548  -73.607143  53.387097   84.266667   
6     Anchorage  1966  -82.387097  -47.285714 -36.419355   57.266667   
7     Anchorage  1967  -93.741935  -50.821429   1.290323   55.266667   
8     Anchorage  1968  -75.580645  -20.310345  17.000000   49.066667   
9     Anchorage  1969 -112.322581  -35.285714  15.677419   79.166667   
10    Anchorage  1970  -87.967742   15.107143  50.354839   59.233333   
11    Anchorage  1971 -121.838710  -27.750000 -47.225806   41.366667   
12    Anchorage  1972 -104.677419  -55.517241 -47.161290   10.400000   
13    Anchorage  1973 -119.419355  -58.571429  -6.354839   51.766667   
14    Anchorage  1974  -99.838710  -65.142857   1.322581   74.600000   
15    Anchorage  1975  -69.548387  -62.642857  -9.483871   35.566667   
16    Anchorage  1976  -54.193548  -62.241379 -15.032258   51.033333   
17    Anchorage  1977   24.354839   35.750000  -2.129032   53.933333   
18    Anchorage  1978  -28.419355    5.214286  24.580645   83.700000   
19    Anchorage  1979  -19.967742  -64.035714  30.322581   71.966667   
20    Anchorage  1980  -61.677419    9.586207  11.290323   86.900000   
21    Anchorage  1981   28.161290  -10.357143  51.935484   72.600000   
22    Anchorage  1982 -100.161290  -54.928571   4.322581   47.200000   
23    Anchorage  1983  -63.516129  -23.500000  26.258065   66.000000   
24    Anchorage  1984  -38.354839  -34.482759  65.322581   78.466667   
25    Anchorage  1985   22.645161  -62.928571  12.548387   23.666667   
26    Anchorage  1986   -7.161290  -18.785714   2.548387   36.500000   
27    Anchorage  1987  -17.548387    0.321429  16.290323   76.866667   
28    Anchorage  1988  -44.516129  -13.965517  27.258065   65.200000   
29    Anchorage  1989 -116.322581  -37.928571   3.064516   83.100000   
...         ...   ...         ...         ...        ...         ...   
1129   Winnipeg  1973  -89.000000  -84.142857  46.741935   91.400000   
1130   Winnipeg  1974 -170.741935 -111.464286 -50.483871   69.566667   
1131   Winnipeg  1975 -108.548387 -108.321429 -56.225806   68.133333   
1132   Winnipeg  1976 -121.387097  -57.965517 -39.096774  127.666667   
1133   Winnipeg  1977 -160.580645  -65.785714  22.741935  160.133333   
1134   Winnipeg  1978 -168.354839 -122.642857 -27.354839   87.133333   
1135   Winnipeg  1979 -178.451613 -167.535714 -50.870968   38.800000   
1136   Winnipeg  1980 -139.096774 -111.551724 -40.258065  161.000000   
1137   Winnipeg  1981  -84.322581  -54.857143  44.741935  118.000000   
1138   Winnipeg  1982 -198.806452  -95.321429 -24.645161   91.100000   
1139   Winnipeg  1983  -77.935484  -60.750000 -12.322581   90.933333   
1140   Winnipeg  1984 -108.483871  -24.586207 -18.709677  141.933333   
1141   Winnipeg  1985 -126.032258 -114.321429  18.354839  138.633333   
1142   Winnipeg  1986  -80.000000 -103.071429   8.451613  101.266667   
1143   Winnipeg  1987  -70.290323  -25.571429  -1.806452  161.433333   
1144   Winnipeg  1988 -131.096774  -98.931034   5.612903  135.433333   
1145   Winnipeg  1989 -100.451613 -131.964286 -55.129032   82.233333   
1146   Winnipeg  1990  -65.741935  -86.285714   3.645161   90.000000   
1147   Winnipeg  1991 -131.354839  -54.035714  -5.225806  134.000000   
1148   Winnipeg  1992  -67.967742  -53.655172  -1.096774   67.666667   
1149   Winnipeg  1993  -95.526316  -94.238095 -19.807692  106.040000   
1150   Winnipeg  1994 -186.741935 -130.892857  28.870968  105.800000   
1151   Winnipeg  1995 -110.258065 -101.000000 -10.677419   65.500000   
1152   Winnipeg  1996 -174.612903 -100.068966 -73.3

From this dataframe, we need a two-dimensional array consisting of weather data values without the city label, and another array consisting of the corresponding city labels. Machine learning models are trained using this format of dataset.

In [2]:
X = data.drop('city', 1).values #remove 'city' column
y = data['city'].values

# Training and Validating

We will try three types of machine learning models: Naive Bayesian, k-Nearest Neighbours, and Support Vector Machine. We will create pipelines to scale each feature's values to lie between 0 and 1 before computing the classifications.

For each model, we will divide the data into training and validation sets. This is how we will find out the accuracy of the trained model - by testing the model on the validation set, which will have never been shown to the model during training.

In [3]:
from sklearn.model_selection import train_test_split

## Naive Bayesian (NB)

In [4]:
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import MinMaxScaler
from sklearn.naive_bayes import GaussianNB

n = 30 #number of iterations
total = 0
for i in range(n):
    X_train, X_valid, y_train, y_valid = train_test_split(X, y)
    bayes_model = make_pipeline(
        MinMaxScaler(),
        GaussianNB()
        )
    bayes_model.fit(X_train, y_train) 
    total += bayes_model.score(X_valid, y_valid)

print("bayes model score: " + str(total/n))

bayes model score: 0.6533333333333334


The bayes model has a score of roughly 65% accuracy, which is terrible. Why is this the case? NB algorithm assumes input variables to be independent, and also assumes input data to be distributed normally. None of these assumptions would hold true for weather data. For example, temperature is highly correlated with snowfall and snow depth. Moreoever, monthly temperatures throughout the decades have no reason to be normally distributed. This explains the model's low score.

## k-Nearest Neighbours (kNN)

In [5]:
from sklearn.neighbors import KNeighborsClassifier

n_nbs = 5 #numer of neighbours to consider
n = 30 #number of iterations
total = 0
for i in range(n):
    X_train, X_valid, y_train, y_valid = train_test_split(X, y)
    knn_model = make_pipeline(
        MinMaxScaler(),
        KNeighborsClassifier(n_neighbors=n_nbs)
        )
    knn_model.fit(X_train, y_train)
    total += knn_model.score(X_valid, y_valid)

print("knn model score: " + str(total/n))

knn model score: 0.6677011494252874


The result is as bad as the NB approach. One guess as to why this algorithm is not effective is because we have mapped the values of all the features to the same scale. Since kNN measures the Euclidian distance between data points in the feature space to determine the class of the prediction input, scaling as mentioned results in all the features contributing approximately the same weight to the distance calcuation. We have no knowledge of which features most significantly affect classification, and it could be the case that some features should be more heavily weighed than others.

## Support Vector Machine (SVM)

In [6]:
from sklearn.svm import SVC

n = 30 #number of iterations
total = 0
for i in range(n):
    X_train, X_valid, y_train, y_valid = train_test_split(X, y)
    svc_model = make_pipeline(
        MinMaxScaler(),
        SVC(kernel='rbf', C=5, gamma='scale', decision_function_shape='ovr')
        )
    svc_model.fit(X_train, y_train)
    total += svc_model.score(X_valid, y_valid)

print("svc model score: " + str(svc_model.score(X_valid, y_valid)))

svc model score: 0.8482758620689655


Our SVM model's accuracy is much better than the other two. SVM is generally more robust to outliers/errors, since it considers all the outliers at the boundary between a pair of classes simultaneously. When more outliers are taken into account simultaneously, they tend to cancel out, which results in a good approximation of the boundary.

# Conclusion

We trained three machine learning models (Naive Bayes, k-Nearest Neighbours, and Support Vector Machine) using a relatively small decades-long weather dataset with 60 features (5 weather measurements * 12 months). The models were trained to predict cities from weather data, then were evaluated for their accuracy. The SVM model had the top accuracy of around 80%, while the other two models had an accuracy of around 66%.

Our SVM model is far from being practically useful. The model can be improved by adding more rows; cities like Denver has fewer data points than the rest of the cities. Perhaps Denver's (and other cities') missing weather information can be researched and added to the table. Another way to improve the model is to add more relevant features to our data. Adding wind velocity and atmospheric pressure, for example, will significantly improve all our models' accuracy.